plutus-core/plutus-ir/src/PlutusIR/Transform/Inline.hs

```diff
@@ -357,59 +353,71 @@ maybeAddSubst body a s n rhs = do
     hints <- view iiHints
     let hinted = shouldInline hints a n
 
-    preUnconditional <- preInlineUnconditional rhs'
-    if preUnconditional
+    if hinted -- if we've been told specifically, then do it right away
     then extendAndDrop (Done $ dupable rhs')
     else do
-        -- See Note [Inlining approach and 'Secrets of the GHC Inliner']
-        postUnconditional <- postInlineUnconditional rhs'
-        if hinted || postUnconditional
+        termIsPure <- checkPurity rhs'
+        preUnconditional <- liftM2 (&&) (termUsedAtMostOnce n) (effectSafe body s n termIsPure)
+        -- similar to the paper, preUnconditional inlining checks that the binder is 'OnceSafe'.
```

Comment:
I don't think our "Safe" and their "Safe" mean the same thing.
Their "Safe" means "don't duplicate work".
Our "Safe" means "don't alter the semantics of the program".

Reply:
Not altering the semantics of the program is a given. Ideally we also want to not duplicate work.

Reply:
Right, but Ziyang is pointing out that we currently use the words differently. And I think the comment here _does_ confuse those senses, because it references the paper's use of "safe" right next to a different use of "safe".

Reply:
Updated.